# American Express Default Prediction Project
#### Mia Belli & Shruti Lakhani - BUS 316

In [1]:
import pandas as pd
import numpy as np
import sklearn.linear_model

In [2]:
data = pd.read_csv('train_data.csv', nrows = 20000)

In [3]:
data

customer_ID         S_2  \
0      0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  2017-03-09   
1      0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  2017-04-07   
2      0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  2017-05-28   
3      0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  2017-06-13   
4      0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  2017-07-16   
...                                                  ...         ...   
19995  00f1aea27e2b2e8504b3c0e3c691e1ed60b16a9d6b5961...  2017-03-03   
19996  00f1aea27e2b2e8504b3c0e3c691e1ed60b16a9d6b5961...  2017-04-03   
19997  00f1aea27e2b2e8504b3c0e3c691e1ed60b16a9d6b5961...  2017-05-03   
19998  00f1aea27e2b2e8504b3c0e3c691e1ed60b16a9d6b5961...  2017-06-03   
19999  00f1aea27e2b2e8504b3c0e3c691e1ed60b16a9d6b5961...  2017-07-03   

            P_2      D_39       B_1       B_2       R_1       S_3      D_41  \
0      0.938469  0.001733  0.008724  1.006838  0.009228  0.124035  0.008771   
1      0.936665  0.005775  0.004923  1.000653  0.006151  0.126750  0.000798   
2      0.954180  0.091505  0.021655  1.009672  0.006815  0.123977  0.007598   
3      0.960384  0.002455  0.013683  1.002700  0.001373  0.117169  0.000685   
4      0.947248  0.002483  0.015193  1.000727  0.007605  0.117325  0.004653   
...         ...       ...       ...       ...       ...       ...       ...   
19995  0.870948  0.004101  0.004307  0.813405  0.001476       NaN  0.005337   
19996  0.875560  0.001806  0.009063  0.814070  0.008061       NaN  0.005790   
19997  0.902034  0.009221  0.007993  0.811671  0.003849       NaN  0.002478   
19998  0.865017  0.006772  0.006316  0.813458  0.009021       NaN  0.007825   
19999  0.910519  0.003323  0.002007  0.813010  0.000483       NaN  0.008578   

            B_3  ...  D_136  D_137  D_138     D_139     D_140     D_141  \
0      0.004709  ...    NaN    NaN    NaN  0.002427  0.003706  0.003818   
1      0.002714  ...    NaN    NaN    NaN  0.003954  0.003167  0.005032   
2      0.009423  ...    NaN    NaN    NaN  0.003269  0.007329  0.000427   
3      0.005531  ...    NaN    NaN    NaN  0.006117  0.004516  0.003200   
4      0.009312  ...    NaN    NaN    NaN  0.003671  0.004946  0.008889   
...         ...  ...    ...    ...    ...       ...       ...       ...   
19995  0.002361  ...    NaN    NaN    NaN  0.008456  0.008546  0.007936   
19996  0.003433  ...    NaN    NaN    NaN  0.004083  0.009451  0.001949   
19997  0.008629  ...    NaN    NaN    NaN  0.006757  0.003446  0.004880   
19998  0.005890  ...    NaN    NaN    NaN  0.009640  0.006659  0.001990   
19999  0.005668  ...    NaN    NaN    NaN  0.005055  0.003387  0.006381   

       D_142     D_143     D_144     D_145  
0        NaN  0.000569  0.000610  0.002674  
1        NaN  0.009576  0.005492  0.009217  
2        NaN  0.003429  0.006986  0.002603  
3        NaN  0.008419  0.006527  0.009600  
4        NaN  0.001670  0.008126  0.009827  
...      ...       ...       ...       ...  
19995    NaN  0.008600  0.002080  0.002491  
19996    NaN  0.007942  0.002523  0.001238  
19997    NaN  0.002150  0.004574  0.006640  
19998    NaN  0.001636  0.007239  0.006323  
19999    NaN  0.009071  0.000452  0.009093  

[20000 rows x 190 columns]

In [4]:
# selecting variables that are only balance and deliquency variables >> because deliquency gives us a variable based off of past behavior
# while balance variables give us a current value that of what the customer owes
df=data[[ 'customer_ID','D_39', 'B_1', 'B_2', 'D_41', 'B_3', 'D_42', 'D_43', 'D_44', 'B_4', 'D_45', 'B_5', 'D_46', 'D_47', 'D_48', 'D_49', 'B_6', 'B_7', 'B_8', 'D_50', 'D_51', 'B_9', 'D_52', 'B_10', 'D_53', 'B_11', 'D_54', 'B_12', 'D_55', 'D_56', 'B_13', 'D_58', 'B_14', 'D_59', 'D_60', 'D_61', 'B_15', 'D_62', 'D_63', 'D_64', 'D_65', 'B_16', 'B_17', 'B_18', 'B_19', 'D_66', 'B_20', 'D_68', 'B_21', 'D_69', 'B_22', 'D_70', 'D_71', 'D_72', 'B_23', 'D_73', 'D_74', 'D_75', 'D_76', 'B_24', 'D_77', 'B_25', 'B_26', 'D_78', 'D_79', 'D_80', 'B_27', 'D_81', 'D_82', 'B_28', 'D_83', 'D_84', 'B_29', 'B_30', 'D_86', 'D_87', 'D_88', 'B_31', 'B_32', 'B_33', 'D_89', 'D_91', 'D_92', 'D_93', 'D_94', 'D_96', 'D_102', 'D_103', 'D_104', 'D_105', 'D_106', 'D_107', 'B_36', 'B_37', 'B_38', 'D_108', 'D_109', 'D_110', 'D_111', 'B_39', 'D_112', 'B_40', 'D_113', 'D_114', 'D_115', 'D_116', 'D_117', 'D_118', 'D_119', 'D_120', 'D_121', 'D_122', 'D_123', 'D_124', 'D_125', 'D_126', 'D_127', 'D_128', 'D_129', 'B_41', 'B_42', 'D_130', 'D_131', 'D_132', 'D_133', 'D_134', 'D_135', 'D_136', 'D_137', 'D_138', 'D_139', 'D_140', 'D_141', 'D_142', 'D_143', 'D_144', 'D_145']]

In [5]:
df

customer_ID      D_39       B_1  \
0      0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  0.001733  0.008724   
1      0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  0.005775  0.004923   
2      0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  0.091505  0.021655   
3      0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  0.002455  0.013683   
4      0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  0.002483  0.015193   
...                                                  ...       ...       ...   
19995  00f1aea27e2b2e8504b3c0e3c691e1ed60b16a9d6b5961...  0.004101  0.004307   
19996  00f1aea27e2b2e8504b3c0e3c691e1ed60b16a9d6b5961...  0.001806  0.009063   
19997  00f1aea27e2b2e8504b3c0e3c691e1ed60b16a9d6b5961...  0.009221  0.007993   
19998  00f1aea27e2b2e8504b3c0e3c691e1ed60b16a9d6b5961...  0.006772  0.006316   
19999  00f1aea27e2b2e8504b3c0e3c691e1ed60b16a9d6b5961...  0.003323  0.002007   

            B_2      D_41       B_3  D_42  D_43      D_44       B_4  ...  \
0      1.006838  0.008771  0.004709   NaN   NaN  0.000630  0.080986  ...   
1      1.000653  0.000798  0.002714   NaN   NaN  0.002526  0.069419  ...   
2      1.009672  0.007598  0.009423   NaN   NaN  0.007605  0.068839  ...   
3      1.002700  0.000685  0.005531   NaN   NaN  0.006406  0.055630  ...   
4      1.000727  0.004653  0.009312   NaN   NaN  0.007731  0.038862  ...   
...         ...       ...       ...   ...   ...       ...       ...  ...   
19995  0.813405  0.005337  0.002361   NaN   NaN  0.006182  0.021971  ...   
19996  0.814070  0.005790  0.003433   NaN   NaN  0.006241  0.019230  ...   
19997  0.811671  0.002478  0.008629   NaN   NaN  0.002818  0.035471  ...   
19998  0.813458  0.007825  0.005890   NaN   NaN  0.003867  0.015147  ...   
19999  0.813010  0.008578  0.005668   NaN   NaN  0.002834  0.027636  ...   

       D_136  D_137  D_138     D_139     D_140     D_141  D_142     D_143  \
0        NaN    NaN    NaN  0.002427  0.003706  0.003818    NaN  0.000569   
1        NaN    NaN    NaN  0.003954  0.003167  0.005032    NaN  0.009576   
2        NaN    NaN    NaN  0.003269  0.007329  0.000427    NaN  0.003429   
3        NaN    NaN    NaN  0.006117  0.004516  0.003200    NaN  0.008419   
4        NaN    NaN    NaN  0.003671  0.004946  0.008889    NaN  0.001670   
...      ...    ...    ...       ...       ...       ...    ...       ...   
19995    NaN    NaN    NaN  0.008456  0.008546  0.007936    NaN  0.008600   
19996    NaN    NaN    NaN  0.004083  0.009451  0.001949    NaN  0.007942   
19997    NaN    NaN    NaN  0.006757  0.003446  0.004880    NaN  0.002150   
19998    NaN    NaN    NaN  0.009640  0.006659  0.001990    NaN  0.001636   
19999    NaN    NaN    NaN  0.005055  0.003387  0.006381    NaN  0.009071   

          D_144     D_145  
0      0.000610  0.002674  
1      0.005492  0.009217  
2      0.006986  0.002603  
3      0.006527  0.009600  
4      0.008126  0.009827  
...         ...       ...  
19995  0.002080  0.002491  
19996  0.002523  0.001238  
19997  0.004574  0.006640  
19998  0.007239  0.006323  
19999  0.000452  0.009093  

[20000 rows x 137 columns]

In [6]:
# dropping columns with null values
df1 = df.dropna(axis='columns')

In [7]:
df1

customer_ID      D_39       B_1  \
0      0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  0.001733  0.008724   
1      0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  0.005775  0.004923   
2      0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  0.091505  0.021655   
3      0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  0.002455  0.013683   
4      0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  0.002483  0.015193   
...                                                  ...       ...       ...   
19995  00f1aea27e2b2e8504b3c0e3c691e1ed60b16a9d6b5961...  0.004101  0.004307   
19996  00f1aea27e2b2e8504b3c0e3c691e1ed60b16a9d6b5961...  0.001806  0.009063   
19997  00f1aea27e2b2e8504b3c0e3c691e1ed60b16a9d6b5961...  0.009221  0.007993   
19998  00f1aea27e2b2e8504b3c0e3c691e1ed60b16a9d6b5961...  0.006772  0.006316   
19999  00f1aea27e2b2e8504b3c0e3c691e1ed60b16a9d6b5961...  0.003323  0.002007   

            B_4       B_5      D_47       B_6       B_7      D_51       B_9  \
0      0.080986  0.170600  0.525351  0.063902  0.059416  1.335856  0.008207   
1      0.069419  0.113239  0.521311  0.065261  0.057744  1.339794  0.008373   
2      0.068839  0.060492  0.524568  0.066982  0.056647  1.337179  0.009355   
3      0.055630  0.166782  0.530929  0.083720  0.049253  1.339909  0.006782   
4      0.038862  0.143630  0.529305  0.075900  0.048918  1.341735  0.000519   
...         ...       ...       ...       ...       ...       ...       ...   
19995  0.021971  0.001197  0.207969  0.199158  0.030578  0.000252  0.007802   
19996  0.019230  0.007820  0.202613  0.158542  0.025895  0.008708  0.005094   
19997  0.035471  0.007346  0.207010  0.191696  0.043366  0.003039  0.002716   
19998  0.015147  0.003382  0.213998  0.176961  0.030651  0.007282  0.005369   
19999  0.027636  0.005206  0.216104  0.213675  0.041360  0.008804  0.002511   

       ...  B_31      B_32      D_92      D_93      D_94      D_96      B_36  \
0      ...     1  0.006626  1.006133  0.003569  0.008871  0.004950  0.009968   
1      ...     1  0.001854  1.005791  0.000571  0.000391  0.003180  0.003921   
2      ...     1  0.008686  1.005801  0.007425  0.009234  0.005433  0.001264   
3      ...     1  0.002478  1.007036  0.000664  0.003200  0.000063  0.002729   
4      ...     1  0.002199  1.002915  0.003079  0.003845  0.000535  0.009998   
...    ...   ...       ...       ...       ...       ...       ...       ...   
19995  ...     1  0.000883  0.004938  0.004410  0.006293  0.007002  0.004829   
19996  ...     1  0.000112  0.002669  0.001101  0.001051  0.004105  0.005805   
19997  ...     1  0.008237  0.009425  0.008736  0.003516  0.004011  0.005973   
19998  ...     1  0.007263  0.000815  0.005635  0.008626  0.002651  0.002118   
19999  ...     1  0.007593  0.003792  0.004553  0.005541  0.004329  0.006891   

           B_37      B_40     D_127  
0      0.004572  0.210060  1.003319  
1      0.004654  0.184093  1.008394  
2      0.019176  0.154837  1.009307  
3      0.011720  0.153939  1.001671  
4      0.017598  0.120717  1.009886  
...         ...       ...       ...  
19995  0.001630  0.034719  0.000159  
19996  0.000408  0.027633  0.001471  
19997  0.004494  0.047069  0.009010  
19998  0.002351  0.022894  0.005144  
19999  0.000378  0.042674  0.008179  

[20000 rows x 36 columns]

In [8]:
cols = df1.columns.values.tolist()
print(cols)

['customer_ID', 'D_39', 'B_1', 'B_4', 'B_5', 'D_47', 'B_6', 'B_7', 'D_51', 'B_9', 'B_10', 'B_11', 'B_12', 'D_58', 'B_14', 'D_60', 'D_63', 'D_65', 'B_18', 'B_21', 'D_71', 'B_23', 'D_75', 'B_24', 'B_28', 'D_86', 'B_31', 'B_32', 'D_92', 'D_93', 'D_94', 'D_96', 'B_36', 'B_37', 'B_40', 'D_127']


In [9]:
# merging the target variable (y) onto the train dataset 
df_target = pd.read_csv('train_labels.csv', nrows = 20000)
df1 = pd.merge(df1, df_target, how = 'inner', on=['customer_ID'])
df1.head()

customer_ID      D_39       B_1  \
0  0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  0.001733  0.008724   
1  0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  0.005775  0.004923   
2  0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  0.091505  0.021655   
3  0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  0.002455  0.013683   
4  0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  0.002483  0.015193   

        B_4       B_5      D_47       B_6       B_7      D_51       B_9  ...  \
0  0.080986  0.170600  0.525351  0.063902  0.059416  1.335856  0.008207  ...   
1  0.069419  0.113239  0.521311  0.065261  0.057744  1.339794  0.008373  ...   
2  0.068839  0.060492  0.524568  0.066982  0.056647  1.337179  0.009355  ...   
3  0.055630  0.166782  0.530929  0.083720  0.049253  1.339909  0.006782  ...   
4  0.038862  0.143630  0.529305  0.075900  0.048918  1.341735  0.000519  ...   

       B_32      D_92      D_93      D_94      D_96      B_36      B_37  \
0  0.006626  1.006133  0.003569  0.008871  0.004950  0.009968  0.004572   
1  0.001854  1.005791  0.000571  0.000391  0.003180  0.003921  0.004654   
2  0.008686  1.005801  0.007425  0.009234  0.005433  0.001264  0.019176   
3  0.002478  1.007036  0.000664  0.003200  0.000063  0.002729  0.011720   
4  0.002199  1.002915  0.003079  0.003845  0.000535  0.009998  0.017598   

       B_40     D_127  target  
0  0.210060  1.003319       0  
1  0.184093  1.008394       0  
2  0.154837  1.009307       0  
3  0.153939  1.001671       0  
4  0.120717  1.009886       0  

[5 rows x 37 columns]

In [16]:
# finidng variables' correlation
X1 = df1[['D_39', 'B_1', 'B_4', 'B_5', 'D_47', 'B_7', 'D_51', 'B_9', 'B_10', 'B_11', 'B_12', 'D_58', 'B_14', 'D_60', 'D_63', 'D_65', 'B_18', 'B_21', 'D_71', 'B_23', 'D_75', 'B_24', 'B_28', 'D_86', 'B_31', 'B_32', 'D_92', 'D_93', 'D_94', 'D_96', 'B_36', 'D_127']]
corr_matrix = X1.corr()
corr_matrix.head()

D_39       B_1       B_4       B_5      D_47       B_7      D_51  \
D_39  1.000000  0.152899  0.145396  0.045425 -0.008673  0.097549  0.009445   
B_1   0.152899  1.000000  0.499060 -0.054334 -0.124669  0.616029 -0.175440   
B_4   0.145396  0.499060  1.000000  0.008855 -0.015555  0.689690 -0.174716   
B_5   0.045425 -0.054334  0.008855  1.000000  0.062050 -0.087138  0.148671   
D_47 -0.008673 -0.124669 -0.015555  0.062050  1.000000 -0.110652  0.202495   

           B_9      B_10      B_11  ...      B_28      D_86      B_31  \
D_39  0.174236 -0.016372  0.146191  ...  0.167594  0.050622 -0.064229   
B_1   0.556417 -0.037742  0.995940  ...  0.349541 -0.042067 -0.014377   
B_4   0.409261 -0.047836  0.500877  ...  0.739864 -0.018787 -0.026073   
B_5  -0.056090  0.008924 -0.047978  ...  0.232499  0.066914  0.009460   
D_47 -0.278619 -0.000735 -0.105250  ...  0.097029  0.050935  0.013096   

          B_32      D_92      D_93      D_94      D_96      B_36     D_127  
D_39 -0.017918  0.039912  0.000917 -0.011453  0.062151 -0.001768  0.031464  
B_1   0.163679 -0.133077 -0.009984 -0.062910 -0.047145  0.039696 -0.146648  
B_4   0.177661 -0.136312  0.023353 -0.061720 -0.014847  0.054675 -0.201832  
B_5  -0.027191  0.171397  0.000098  0.025391  0.231721 -0.003060  0.132829  
D_47 -0.055762  0.053595  0.050194  0.083399  0.050203  0.010057  0.140877  

[5 rows x 31 columns]

In [18]:
upper_tri = corr_matrix.where(np.triu(np.ones(corr_matrix.shape),k=1).astype(np.bool))
print(upper_tri)

       D_39       B_1       B_4       B_5      D_47       B_7      D_51  \
D_39    NaN  0.152899  0.145396  0.045425 -0.008673  0.097549  0.009445   
B_1     NaN       NaN  0.499060 -0.054334 -0.124669  0.616029 -0.175440   
B_4     NaN       NaN       NaN  0.008855 -0.015555  0.689690 -0.174716   
B_5     NaN       NaN       NaN       NaN  0.062050 -0.087138  0.148671   
D_47    NaN       NaN       NaN       NaN       NaN -0.110652  0.202495   
B_7     NaN       NaN       NaN       NaN       NaN       NaN -0.254214   
D_51    NaN       NaN       NaN       NaN       NaN       NaN       NaN   
B_9     NaN       NaN       NaN       NaN       NaN       NaN       NaN   
B_10    NaN       NaN       NaN       NaN       NaN       NaN       NaN   
B_11    NaN       NaN       NaN       NaN       NaN       NaN       NaN   
B_12    NaN       NaN       NaN       NaN       NaN       NaN       NaN   
D_58    NaN       NaN       NaN       NaN       NaN       NaN       NaN   
B_14    NaN       NaN    

C:\Users\belli\AppData\Local\Temp\ipykernel_8712\1735788728.py:1: DeprecationWarning: `np.bool` is a deprecated alias for the builtin `bool`. To silence this warning, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  upper_tri = corr_matrix.where(np.triu(np.ones(corr_matrix.shape),k=1).astype(np.bool))


In [43]:
# finding variables with a correlation greater than 72%
highest = [column for column in upper_tri.columns if any(upper_tri[column] > 0.72)]
print(highest)

['B_11', 'D_58', 'B_14', 'B_23', 'D_75', 'B_28']


In [32]:
# selecting 2000 rows to use the model on 
df1.sample(n = 2000)

customer_ID      D_39       B_1  \
2980   00236b7198b9c9a91cae37603f0f2c50d9c6a089788a4c...  0.060602  0.020250   
14340  00adcfa58b09501bc9de3a0f68c1d260fe15a13d7b2dae...  0.361703  0.061064   
3397   0029dc52de61302d1db0a2bc4a2649c0cfe3287b493db3...  0.442551  0.888119   
6588   0051111590b08c02df1bb1c375838df6113d2ed0364800...  0.006950  0.001253   
14642  00b040086eefd0f08b24809fbe7263da52f52e6e031e10...  0.003570  0.014947   
...                                                  ...       ...       ...   
3276   002890a84241e45593a3fe6288836eb5a5f128eeaef7fc...  0.303772  0.059161   
16076  00c1a572343e0e1466d32a366305a0dc44c1588f7c09df...  0.265810  0.028148   
18639  00dfc371133228ffa916fcd1ed170b5a541152c8a746c0...  0.031554  0.101452   
6753   005382536094338c2a294250bdfcbfd333781dd093a48b...  0.300027  0.058362   
3629   002cbbf6cb777d7789f9821d4edcffe5b9b607594fc53b...  0.446402  0.061089   

            B_4       B_5      D_47       B_6       B_7      D_51       B_9  \
2980   0.044262  0.094222  0.306635  0.137202  0.035108  0.340942  0.028326   
14340  0.131921  0.020742  0.299566  0.081123  0.137855  0.004671  0.169795   
3397   0.276530  0.024201  0.554792  0.011658  0.631018  0.002071  0.609401   
6588   0.007375  0.007748  0.589156  0.389419  0.012161  0.006793  0.002914   
14642  0.032908  0.061952  0.773429  0.223210  0.022417  0.340798  0.002255   
...         ...       ...       ...       ...       ...       ...       ...   
3276   0.045922  0.126613  0.347066  0.185290  0.052586  0.668365  0.188301   
16076  0.047473  0.004122  0.601927  0.210026  0.039592  0.338211  0.087587   
18639  0.167048  0.006015  0.100362  0.008519  0.437994  0.002632  0.566780   
6753   0.002619  0.032883  0.089829  0.211558  0.054799  0.004352  0.134118   
3629   0.086800  0.087774  0.434997  0.102365  0.099790  0.000421  0.078372   

       ...      B_32      D_92      D_93      D_94      D_96      B_36  \
2980   ...  0.001718  0.008257  0.001968  0.002151  0.000769  0.003505   
14340  ...  0.007105  0.002387  0.005739  0.006676  0.000613  0.003888   
3397   ...  0.000843  0.001543  0.000062  0.008078  0.001062  0.003978   
6588   ...  0.004894  0.005366  0.000118  0.005703  0.009153  0.000574   
14642  ...  0.002649  0.005848  0.007370  0.008653  0.002196  0.004067   
...    ...       ...       ...       ...       ...       ...       ...   
3276   ...  0.000633  1.004529  0.002748  0.005202  1.002451  0.003044   
16076  ...  0.007072  0.002660  0.000722  0.001887  0.000540  0.004895   
18639  ...  0.003578  0.009047  0.009109  0.001575  0.004420  0.004156   
6753   ...  0.003781  0.006981  0.001494  0.004205  0.004160  0.003086   
3629   ...  0.008448  0.004778  0.009072  0.001733  0.004558  0.003708   

           B_37      B_40     D_127  target  
2980   0.025568  0.075551  0.005925       0  
14340  0.062666  0.177617  0.009752       0  
3397   0.890437  0.368008  0.001485       0  
6588   0.006725  0.011056  0.001778       0  
14642  0.017459  0.016887  0.003510       0  
...         ...       ...       ...     ...  
3276   0.057747  0.015253  0.009154       0  
16076  0.027271  0.040636  0.001999       0  
18639  0.104230  0.413652  0.008559       1  
6753   0.054945  0.009947  0.006103       0  
3629   0.070752  0.094649  0.000880       0  

[2000 rows x 37 columns]

In [41]:
# X variables and y variable (target)
X = df1[['B_11', 'D_58', 'B_14', 'B_23', 'D_75', 'B_28']]
y = df1.target
X.shape

# split training set and testing set 
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.25)

# Running Logistic Regression model
model = sklearn.linear_model.LogisticRegression()
model.fit(X_train,y_train)
y_pred = model.predict(X_test)
print(y_pred)

[0 0 1 ... 1 0 0]


In [42]:
# Accuracy of Model
from sklearn.metrics import classification_report
matrix = classification_report(y_test, y_pred)
print(matrix)

              precision    recall  f1-score   support

           0       0.82      0.92      0.87      3691
           1       0.65      0.43      0.51      1309

    accuracy                           0.79      5000
   macro avg       0.73      0.67      0.69      5000
weighted avg       0.77      0.79      0.77      5000



# Our Process

### Why did we decided to only use (D_* ) Delinquency variables & (B_* ) Balance variables?

For the spending variables, we thought those would be unneccessary because we have the balance variables that tell us how much the customers owe after spending. We got rid of the payment variables for the same reason because balance variables will tell us the end result of what a customer needs to pay still. Risk variables could have been helpful for a prediction but delinquency variables seemed like the stronger choice for the prediction model because it tells us a history of delinquency and what could lead to a customer not filling a payment. For example, 'D_58' could have been the percentage of payments that weren't paid on time within the last 5 years. With 190 different variables and over 5.5 million customers, we needed to slim down most of the variables in order to process the program.

### How else did we clean the data?

We started by weeding out the unwanted variables (S, P, and R). Then, deleted columns with any null values because we wanted columns that had the most information to give for the prediction. Afterwards, we merged the train_labels csv file to the training dataset because we needed the 'target' variable (y) that described whether a default event would or would not occur for the predictive model. Once finding the variables with the highest correlation, we ran a logistic regression to see if a customer paid or didn't pay in which target = 1 meant they don't pay within the 120 days and target = 0 meant they do pay. 
Note: We didn't have to encode any of the categorical variables because the correlation for the variables didn't label any categorical variables as variables with the highest correlation.

# Our Conclusions

We conclude that 0.65 of the customers are probable to default payments in the future with the accuracy of 0.79. The reasons behind default payments could be financial difficulties faced by the customers also because of the overspending nature of the customers or maybe the customer’s account went through a theft or a fraud. This could be the most probable reasons of customer default payments. To find prevention from default payments American Express should set appropriate credit limits or while offering a high credit limit to the customer they should sign an agreement saying in exchange to increase the credit limit they should keep a high valuable assets from the customer. The second way can be they can sign a petition saying if the customer fails to make the payment they can produce to legal actions. In order to prevent default payments in future. Which will create a security for the bank if the customer doesn’t make the payment on time. This are all the possible ways that AMEX can use to prevent default payments from the customers.